// @umbraco-cms/backoffice/dist-cms/libs/context-api/consume/context-consume.decorator.js
/*
 * Portions of this code are adapted from @lit/context
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Original source: https://github.com/lit/lit/tree/main/packages/context
 *
 * @license BSD-3-Clause
 */
import { UmbContextConsumerController } from './context-consumer.controller.js';
/**
 * A property decorator that adds an UmbContextConsumerController to the component
 * which will try and retrieve a value for the property via the Umbraco Context API.
 *
 * This decorator supports both modern "standard" decorators (Stage 3 TC39 proposal) and
 * legacy TypeScript experimental decorators for backward compatibility.
 * @param {UmbConsumeOptions} options Configuration object containing context, callback, and subscribe options
 * @example
 * ```ts
 * import {consumeContext} from '../index.js';
 * import {UMB_WORKSPACE_CONTEXT} from './workspace.context-token.js';
 *
 * class MyElement extends UmbLitElement {
 *   // Standard decorators (with 'accessor' keyword) - Modern approach
 *   @consumeContext({context: UMB_WORKSPACE_CONTEXT})
 *   accessor workspaceContext?: UmbWorkspaceContext;
 *
 *   // Legacy decorators (without 'accessor') - Works with @state/@property
 *   @consumeContext({context: UMB_USER_CONTEXT, subscribe: false})
 *   @state()
 *   currentUser?: UmbUserContext;
 * }
 * ```
 * @returns {ConsumeDecorator<ResultType>} A property decorator function
 */
export function consumeContext(options) {
    const { context, callback, subscribe = true } = options;
    return ((protoOrTarget, nameOrContext) => {
        if (typeof nameOrContext === 'object') {
            setupStandardDecorator(protoOrTarget, nameOrContext, context, callback, subscribe);
            return;
        }
        setupLegacyDecorator(protoOrTarget, nameOrContext, context, callback, subscribe);
    });
}
/**
 * Sets up a standard decorator (Stage 3 TC39 proposal) for auto-accessors.
 * This branch is used when decorating with the 'accessor' keyword.
 * Example: @consumeContext({context: TOKEN}) accessor myProp?: Type;
 *
 * The decorator receives a ClassAccessorDecoratorContext object which provides
 * addInitializer() to run code during class construction.
 *
 * This is the modern, standardized decorator API that will be the standard
 * when Lit 4.x is released.
 *
 * Note: Standard decorators currently don't work with @state()/@property()
 * decorators, which is why we still need the legacy branch.
 * @param protoOrTarget
 * @param decoratorContext
 * @param context
 * @param callback
 * @param subscribe
 */
function setupStandardDecorator(protoOrTarget, decoratorContext, context, callback, subscribe) {
    if (!('addInitializer' in decoratorContext)) {
        console.warn('@consumeContext decorator: Standard decorator context does not support addInitializer. ' +
            'This should not happen with modern decorators.');
        return;
    }
    decoratorContext.addInitializer(function () {
        queueMicrotask(() => {
            if (subscribe) {
                // Continuous subscription - stays active and updates property on context changes
                new UmbContextConsumerController(this, context, (value) => {
                    protoOrTarget.set.call(this, value);
                    callback?.(value);
                });
            }
            else {
                // One-time consumption - uses asPromise() to get the value once and then cleans up
                const controller = new UmbContextConsumerController(this, context, callback);
                controller.asPromise().then((value) => {
                    protoOrTarget.set.call(this, value);
                });
            }
        });
    });
}
/**
 * Sets up a legacy decorator (TypeScript experimental) for regular properties.
 * This branch is used when decorating without the 'accessor' keyword.
 * Example: @consumeContext({context: TOKEN}) @state() myProp?: Type;
 *
 * The decorator receives:
 * - protoOrTarget: The class prototype
 * - propertyKey: The property name (string)
 *
 * This is the older TypeScript experimental decorator API, still widely used
 * in Umbraco because it works with @state() and @property() decorators.
 * The 'accessor' keyword is not compatible with these decorators yet.
 *
 * We support three initialization strategies:
 * 1. addInitializer (if available, e.g., on LitElement classes)
 * 2. hostConnected wrapper (for UmbController classes)
 * 3. Warning (if neither is available)
 * @param protoOrTarget
 * @param propertyKey
 * @param context
 * @param callback
 * @param subscribe
 */
function setupLegacyDecorator(protoOrTarget, propertyKey, context, callback, subscribe) {
    const constructor = protoOrTarget.constructor;
    // Strategy 1: Use addInitializer if available (LitElement classes)
    if (constructor.addInitializer) {
        constructor.addInitializer((element) => {
            queueMicrotask(() => {
                if (subscribe) {
                    // Continuous subscription
                    new UmbContextConsumerController(element, context, (value) => {
                        element[propertyKey] = value;
                        callback?.(value);
                    });
                }
                else {
                    // One-time consumption using asPromise()
                    const controller = new UmbContextConsumerController(element, context, callback);
                    controller.asPromise().then((value) => {
                        element[propertyKey] = value;
                    });
                }
            });
        });
        return;
    }
    // Strategy 2: Wrap hostConnected for UmbController classes without addInitializer
    if ('hostConnected' in protoOrTarget && typeof protoOrTarget.hostConnected === 'function') {
        const originalHostConnected = protoOrTarget.hostConnected;
        protoOrTarget.hostConnected = function () {
            // Set up consumer once, using a flag to prevent multiple setups
            if (!this.__consumeControllers) {
                this.__consumeControllers = new Map();
            }
            if (!this.__consumeControllers.has(propertyKey)) {
                if (subscribe) {
                    // Continuous subscription
                    const controller = new UmbContextConsumerController(this, context, (value) => {
                        this[propertyKey] = value;
                        callback?.(value);
                    });
                    this.__consumeControllers.set(propertyKey, controller);
                }
                else {
                    // One-time consumption using asPromise()
                    const controller = new UmbContextConsumerController(this, context, callback);
                    controller.asPromise().then((value) => {
                        this[propertyKey] = value;
                    });
                    // Don't store in map since it cleans itself up
                }
            }
            // Call original hostConnected if it exists
            originalHostConnected?.call(this);
        };
        return;
    }
    // Strategy 3: No supported initialization method available
    console.warn(`@consumeContext applied to ${constructor.name}.${propertyKey} but neither addInitializer nor hostConnected is available. ` +
        `Make sure the class extends UmbLitElement, UmbControllerBase, or implements UmbController with hostConnected.`);
}
