// @umbraco-cms/backoffice/dist-cms/libs/context-api/provide/context-provide.decorator.js
/*
 * Portions of this code are adapted from @lit/context
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Original source: https://github.com/lit/lit/tree/main/packages/context
 *
 * @license BSD-3-Clause
 */
import { UmbContextProviderController } from './context-provider.controller.js';
/**
 * A property decorator that creates an UmbContextProviderController to provide
 * a context value to child elements via the Umbraco Context API.
 *
 * This decorator supports both modern "standard" decorators (Stage 3 TC39 proposal) and
 * legacy TypeScript experimental decorators for backward compatibility.
 *
 * The provider is created once during initialization with the property's initial value.
 * To update the provided value dynamically, keep a state inside the provided context instance
 * and update that state as needed. The context instance itself should remain the same.
 * You can use any of the Umb{*}State classes.
 * @param {UmbProvideOptions} options Configuration object containing the context token
 * @example
 * ```ts
 * import {provideContext} from '../index.js';
 * import {UMB_WORKSPACE_CONTEXT} from './workspace.context-token.js';
 *
 * class MyWorkspaceElement extends UmbLitElement {
 *   // Standard decorators - requires 'accessor' keyword
 *   @provideContext({context: UMB_WORKSPACE_CONTEXT})
 *   accessor workspaceContext = new UmbWorkspaceContext(this);
 *
 *   // Legacy decorators - works without 'accessor'
 *   @provideContext({context: UMB_WORKSPACE_CONTEXT})
 *   workspaceContext = new UmbWorkspaceContext(this);
 * }
 * ```
 * @example
 * ```ts
 * // For dynamic updates, store the state inside the context instance
 * class MyContext extends UmbControllerBase {
 *   someProperty = new UmbStringState('initial value');
 * }
 *
 * class MyElement extends UmbLitElement {
 *   @provideContext({context: MY_CONTEXT})
 *   private _myContext = new MyContext(this);
 *
 *   updateValue(newValue: string) {
 *     this._myContext.someProperty.setValue(newValue);
 *   }
 * }
 * ```
 * @returns {ProvideDecorator<InstanceType>} A property decorator function
 */
export function provideContext(options) {
    const { context } = options;
    return ((protoOrTarget, nameOrContext) => {
        if (typeof nameOrContext === 'object') {
            return setupStandardDecorator(protoOrTarget, context);
        }
        setupLegacyDecorator(protoOrTarget, nameOrContext, context);
    });
}
/**
 * Sets up a standard decorator (Stage 3 TC39 proposal) for auto-accessors.
 * This branch is used when decorating with the 'accessor' keyword.
 * Example: @provideContext({context: TOKEN}) accessor myProp = new MyContext();
 *
 * The decorator receives a ClassAccessorDecoratorContext object and returns
 * an accessor descriptor that intercepts the property initialization.
 *
 * This is the modern, standardized decorator API that will be the standard
 * when Lit 4.x is released.
 *
 * Note: Standard decorators currently don't work with @state()/@property()
 * decorators, which is why we still need the legacy branch.
 * @param protoOrTarget
 * @param context
 */
function setupStandardDecorator(protoOrTarget, context) {
    return {
        get() {
            return protoOrTarget.get.call(this);
        },
        set(value) {
            return protoOrTarget.set.call(this, value);
        },
        init(value) {
            // Defer controller creation to avoid timing issues with private fields
            queueMicrotask(() => {
                new UmbContextProviderController(this, context, value);
            });
            return value;
        },
    };
}
/**
 * Sets up a legacy decorator (TypeScript experimental) for regular properties.
 * This branch is used when decorating without the 'accessor' keyword.
 * Example: @provideContext({context: TOKEN}) myProp = new MyContext();
 *
 * The decorator receives:
 * - protoOrTarget: The class prototype
 * - propertyKey: The property name (string)
 *
 * This is the older TypeScript experimental decorator API, still widely used
 * in Umbraco because it works with @state() and @property() decorators.
 * The 'accessor' keyword is not compatible with these decorators yet.
 *
 * We support three initialization strategies:
 * 1. addInitializer (if available, e.g., on LitElement classes)
 * 2. hostConnected wrapper (for UmbController classes)
 * 3. Warning (if neither is available)
 * @param protoOrTarget
 * @param propertyKey
 * @param context
 */
function setupLegacyDecorator(protoOrTarget, propertyKey, context) {
    const constructor = protoOrTarget.constructor;
    // Strategy 1: Use addInitializer if available (LitElement classes)
    if (constructor.addInitializer) {
        constructor.addInitializer((element) => {
            // Defer controller creation to avoid timing issues with private fields
            queueMicrotask(() => {
                const initialValue = element[propertyKey];
                new UmbContextProviderController(element, context, initialValue);
            });
        });
        return;
    }
    // Strategy 2: Wrap hostConnected for UmbController classes without addInitializer
    if ('hostConnected' in protoOrTarget && typeof protoOrTarget.hostConnected === 'function') {
        const originalHostConnected = protoOrTarget.hostConnected;
        protoOrTarget.hostConnected = function () {
            // Set up provider once, using a flag to prevent multiple setups
            if (!this.__provideControllers) {
                this.__provideControllers = new Map();
            }
            if (!this.__provideControllers.has(propertyKey)) {
                const initialValue = this[propertyKey];
                new UmbContextProviderController(this, context, initialValue);
                // Mark as set up to prevent duplicate providers
                this.__provideControllers.set(propertyKey, true);
            }
            // Call original hostConnected if it exists
            originalHostConnected?.call(this);
        };
        return;
    }
    // Strategy 3: No supported initialization method available
    console.warn(`@provideContext applied to ${constructor.name}.${propertyKey} but neither addInitializer nor hostConnected is available. ` +
        `Make sure the class extends UmbLitElement, UmbControllerBase, or implements UmbController with hostConnected.`);
}
