// @umbraco-cms/backoffice/dist-cms/external/openid/src/index.js
/**
 * @deprecated This package is no longer used by Umbraco auth. Scheduled for removal in Umbraco 19.
 * The Umbraco backoffice now uses cookie-based authentication with a minimal PKCE client.
 * Data classes (TokenResponse, AuthorizationRequest, etc.) remain functional for compatibility.
 * Handler classes (RedirectRequestHandler, BaseTokenRequestHandler, etc.) reject with an error
 * because the underlying OAuth operations are no longer possible with cookie-based auth.
 */
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class TokenResponse {
    constructor(response) {
        this.accessToken = response.access_token;
        this.tokenType = response.token_type || 'bearer';
        if (response.expires_in) {
            this.expiresIn = parseInt(response.expires_in, 10);
        }
        this.refreshToken = response.refresh_token;
        this.scope = response.scope;
        this.idToken = response.id_token;
        this.issuedAt = response.issued_at || Math.round(Date.now() / 1000);
    }
    toJson() {
        return {
            access_token: this.accessToken,
            id_token: this.idToken,
            refresh_token: this.refreshToken,
            scope: this.scope,
            token_type: this.tokenType,
            issued_at: this.issuedAt,
            expires_in: this.expiresIn?.toString(),
        };
    }
    isValid(buffer = 0) {
        if (this.expiresIn) {
            const now = Math.round(Date.now() / 1000);
            return now < this.issuedAt + this.expiresIn + buffer;
        }
        return true;
    }
}
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class TokenError {
    constructor(tokenError) {
        this.error = tokenError.error;
        this.errorDescription = tokenError.error_description;
        this.errorUri = tokenError.error_uri;
    }
    toJson() {
        return {
            error: this.error,
            error_description: this.errorDescription,
            error_uri: this.errorUri,
        };
    }
}
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class AuthorizationServiceConfiguration {
    constructor(request) {
        this.authorizationEndpoint = request.authorization_endpoint;
        this.tokenEndpoint = request.token_endpoint;
        this.revocationEndpoint = request.revocation_endpoint;
        this.userInfoEndpoint = request.userinfo_endpoint;
        this.endSessionEndpoint = request.end_session_endpoint;
    }
    toJson() {
        return {
            authorization_endpoint: this.authorizationEndpoint,
            token_endpoint: this.tokenEndpoint,
            revocation_endpoint: this.revocationEndpoint,
            end_session_endpoint: this.endSessionEndpoint,
            userinfo_endpoint: this.userInfoEndpoint,
        };
    }
}
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class AuthorizationRequest {
    static { this.RESPONSE_TYPE_TOKEN = 'token'; }
    static { this.RESPONSE_TYPE_CODE = 'code'; }
    constructor(request) {
        this.clientId = request.client_id;
        this.redirectUri = request.redirect_uri;
        this.scope = request.scope;
        this.responseType = request.response_type || AuthorizationRequest.RESPONSE_TYPE_CODE;
        this.state = request.state || '';
        this.extras = request.extras;
        this.internal = request.internal;
    }
    setupCodeVerifier() {
        return Promise.resolve();
    }
    toJson() {
        return Promise.resolve({
            response_type: this.responseType,
            client_id: this.clientId,
            redirect_uri: this.redirectUri,
            scope: this.scope,
            state: this.state,
            extras: this.extras,
            internal: this.internal,
        });
    }
}
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class AuthorizationResponse {
    constructor(response) {
        this.code = response.code;
        this.state = response.state;
    }
}
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class AuthorizationError {
    constructor(error) {
        this.error = error.error;
        this.errorDescription = error.error_description;
        this.errorUri = error.error_uri;
        this.state = error.state;
    }
}
// Stub exports for classes/functions that were previously exported but are no longer functional
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export const GRANT_TYPE_AUTHORIZATION_CODE = 'authorization_code';
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export const GRANT_TYPE_REFRESH_TOKEN = 'refresh_token';
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class AuthorizationNotifier {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    setAuthorizationListener(_listener) {
        // no-op
    }
}
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class LocalStorageBackend {
    async getItem(name) {
        return localStorage.getItem(name);
    }
    async setItem(name, value) {
        localStorage.setItem(name, value);
    }
    async removeItem(name) {
        localStorage.removeItem(name);
    }
}
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class BasicQueryStringUtils {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parse(_input, _useHash) {
        return {};
    }
    stringify(input) {
        return new URLSearchParams(input).toString();
    }
}
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class FetchRequestor {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    xhr(_settings) {
        return Promise.reject(new Error('FetchRequestor is deprecated'));
    }
}
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class RedirectRequestHandler {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    constructor(..._args) {
        // no-op
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    setAuthorizationNotifier(_notifier) {
        // no-op
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    performAuthorizationRequest(_config, _request) {
        return Promise.reject(new Error('RedirectRequestHandler is deprecated'));
    }
    completeAuthorizationRequestIfPossible() {
        return Promise.resolve(null);
    }
    cleanupStaleAuthorizationData() {
        return Promise.resolve();
    }
}
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class BaseTokenRequestHandler {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    constructor(..._args) {
        // no-op
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    performTokenRequest(_config, _request) {
        return Promise.reject(new Error('BaseTokenRequestHandler is deprecated'));
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    performRevokeTokenRequest(_config, _request) {
        return Promise.reject(new Error('BaseTokenRequestHandler is deprecated'));
    }
}
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class TokenRequest {
    constructor(request) {
        this.clientId = request.client_id;
        this.redirectUri = request.redirect_uri;
        this.grantType = request.grant_type;
        this.code = request.code;
        this.refreshToken = request.refresh_token;
        this.extras = request.extras;
    }
}
/**
 * @deprecated Scheduled for removal in Umbraco 19.
 */
export class RevokeTokenRequest {
    constructor(request) {
        this.token = request.token;
        this.tokenTypeHint = request.token_type_hint;
        this.clientId = request.client_id;
    }
}
