// @umbraco-cms/backoffice/dist-cms/apps/backoffice/components/backoffice-main.element.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { UMB_BACKOFFICE_CONTEXT } from '../backoffice.context.js';
import { createExtensionApi, createExtensionElement } from '../../../libs/extension-api/index.js';
import { css, customElement, html, state } from '../../../external/lit/index.js';
import { UmbDefaultSectionContext, UMB_SECTION_PATH_PATTERN } from '../../../packages/core/section/index.js';
import { UmbLitElement } from '../../../packages/core/lit-element/index.js';
let UmbBackofficeMainElement = class UmbBackofficeMainElement extends UmbLitElement {
    #backofficeContext;
    constructor() {
        super();
        this._routes = [];
        this._sections = [];
        this.consumeContext(UMB_BACKOFFICE_CONTEXT, (_instance) => {
            this.#backofficeContext = _instance;
            this._observeBackoffice();
        });
    }
    async _observeBackoffice() {
        if (this.#backofficeContext) {
            this.observe(this.#backofficeContext.allowedSections, (sections) => {
                this._sections = sections.map((section) => section.manifest).filter((section) => section !== undefined);
                this.#createRoutes();
            }, 'observeAllowedSections');
        }
    }
    #createRoutes() {
        if (!this._sections)
            return;
        // TODO: Refactor this for re-use across the app where the routes are re-generated at any time.
        const newRoutes = this._sections.map((manifest) => {
            return {
                path: UMB_SECTION_PATH_PATTERN.generateLocal({ sectionName: manifest.meta.pathname }),
                component: () => createExtensionElement(manifest, 'umb-section-default'),
                setup: async (component) => {
                    if (component) {
                        const element = component;
                        element.manifest = manifest;
                        const api = await createExtensionApi(element, manifest, [], UmbDefaultSectionContext);
                        if (api) {
                            api.manifest = manifest;
                        }
                    }
                    this.#backofficeContext?.setActiveSectionAlias(manifest.alias);
                },
            };
        });
        if (newRoutes.length > 0) {
            newRoutes.push({
                path: '',
                pathMatch: 'full',
                awaitStability: true,
                redirectTo: newRoutes[0].path,
            });
            newRoutes.push({
                path: `**`,
                component: async () => (await import('../../../packages/core/router/index.js')).UmbRouteNotFoundElement,
            });
        }
        this._routes = newRoutes;
    }
    render() {
        if (!this._routes.length)
            return;
        return html `<umb-router-slot .routes=${this._routes}></umb-router-slot>`;
    }
    static { this.styles = [
        css `
			:host {
				display: block;
				background-color: var(--uui-color-background);
				width: 100%;
				height: calc(
					100% - 60px
				); /* 60 => top header height, TODO: Make sure this comes from somewhere so it is maintainable and eventually responsive. */
			}
		`,
    ]; }
};
__decorate([
    state()
], UmbBackofficeMainElement.prototype, "_routes", void 0);
__decorate([
    state()
], UmbBackofficeMainElement.prototype, "_sections", void 0);
UmbBackofficeMainElement = __decorate([
    customElement('umb-backoffice-main')
], UmbBackofficeMainElement);
export { UmbBackofficeMainElement };
