// @umbraco-cms/backoffice/dist-cms/apps/backoffice/backoffice.context.js
import { tryExecute } from '../../packages/core/resources/index.js';
import { umbExtensionsRegistry } from '../../packages/core/extension-registry/index.js';
import { ServerService } from '../../packages/core/backend-api/index.js';
import { UmbBasicState, UmbStringState } from '../../libs/observable-api/index.js';
import { UmbContextBase } from '../../libs/class-api/index.js';
import { UmbContextToken } from '../../libs/context-api/index.js';
import { UmbExtensionsManifestInitializer } from '../../libs/extension-api/index.js';
import { UmbSysinfoRepository } from '../../packages/sysinfo/index.js';
import { UMB_AUTH_CONTEXT } from '../../packages/core/auth/index.js';
import { UMB_CURRENT_USER_CONTEXT } from '../../packages/user/current-user/index.js';
export class UmbBackofficeContext extends UmbContextBase {
    #activeSectionAlias;
    // TODO: We need a class array state:
    #allowedSections;
    #version;
    constructor(host) {
        super(host, UMB_BACKOFFICE_CONTEXT);
        this.#activeSectionAlias = new UmbStringState(undefined);
        this.activeSectionAlias = this.#activeSectionAlias.asObservable();
        // TODO: We need a class array state:
        this.#allowedSections = new UmbBasicState([]);
        this.allowedSections = this.#allowedSections.asObservable();
        this.#version = new UmbStringState(undefined);
        this.version = this.#version.asObservable();
        // TODO: We need to ensure this request is called every time the user logs in, but this should be done somewhere across the app and not here [JOV]
        this.consumeContext(UMB_AUTH_CONTEXT, (authContext) => {
            this.observe(authContext?.isAuthorized, (isAuthorized) => {
                if (!isAuthorized)
                    return;
                this.#getVersion();
            });
        });
        this.consumeContext(UMB_CURRENT_USER_CONTEXT, (userContext) => {
            this.observe(userContext?.allowedSections, (allowedSections) => {
                if (!allowedSections)
                    return;
                // TODO: Please be aware that we re-initialize this initializer based on user permissions. I suggest we should solve this specific case should be improved by the ability to change the filter [NL]
                new UmbExtensionsManifestInitializer(this, umbExtensionsRegistry, 'section', (manifest) => allowedSections.includes(manifest.alias), async (sections) => {
                    this.#allowedSections.setValue(sections);
                }, 'umbAllowedSectionsManifestInitializer');
            }, 'umbAllowedSectionsObserver');
        });
    }
    async #getVersion() {
        const { data } = await tryExecute(this._host, ServerService.getServerInformation(), { disableNotifications: true });
        if (!data)
            return;
        // A quick semver parser (to remove the unwanted bits) [LK]
        // https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const [semVer, major, minor, patch, prerelease, buildmetadata] = data.version.match(/^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/) ?? [];
        const version = [major, minor, patch].join('.') + (prerelease ? `-${prerelease}` : '');
        this.#version.setValue(version);
    }
    setActiveSectionAlias(alias) {
        this.#activeSectionAlias.setValue(alias);
    }
    async serverUpgradeCheck() {
        const version = await this.observe(this.version)
            .asPromise()
            .catch(() => null);
        if (!version)
            return null;
        const repository = new UmbSysinfoRepository(this);
        return repository.serverUpgradeCheck(version);
    }
}
export const UMB_BACKOFFICE_CONTEXT = new UmbContextToken('UmbBackofficeContext');
